### PR TITLE
fix: write back only imported tables whose content actually changed

### LINKED ISSUE #530: [BUG] interactive .save --force rewrites an unchanged CSV import after only TEMP-table changes
## Trigger Conditions
- Published `v0.22.0` binary
- Interactive `.save --force`
- Imported CSV source is unchanged
- Only a `TEMP` table was created and modified

## Reproduction
```bash
cp testdata/sample.csv /tmp/temp_only.csv
printf 'CREATE TEMP TABLE scratch(id INTEGER);\nINSERT INTO scratch VALUES (1);\n.save --force\n' | sqly /tmp/temp_only.csv
```

## Actual Behavior
`sqly` prints `Saved temp_only to /tmp/temp_only.csv` and rewrites the source bytes even though the imported table data never changed.

## Expected Behavior
No save should happen because no file-backed imported table changed.

## Suggested Regression Coverage
- Add an e2e test for `.save --force` after temp-table-only changes.
- Assert that the imported CSV file bytes stay unchanged and no save message is emitted.

### LINKED ISSUE #531: [BUG] interactive .save DIR rewrites an unchanged CSV export after only TEMP-table changes
## Trigger Conditions
- Published `v0.22.0` binary
- Interactive `.save DIR`
- Imported CSV source is unchanged
- Only a `TEMP` table was created and modified

## Reproduction
```bash
cp testdata/sample.csv /tmp/temp_only_dir.csv
rm -rf /tmp/temp-only-out
printf 'CREATE TEMP TABLE scratch(id INTEGER);\nINSERT INTO scratch VALUES (1);\n.save /tmp/temp-only-out\n' | sqly /tmp/temp_only_dir.csv
```

## Actual Behavior
`sqly` writes `/tmp/temp-only-out/temp_only_dir.csv` even though the imported table data never changed.

## Expected Behavior
No export should be written because no file-backed imported table changed.

## Suggested Regression Coverage
- Add an e2e test for `.save DIR` after temp-table-only changes.
- Assert that no output file is created when only temp tables changed.

### LINKED ISSUE #532: [BUG] interactive .save --force fails on an unchanged JSONL import after only TEMP-table changes
## Trigger Conditions
- Published `v0.22.0` binary
- Interactive `.save --force`
- Imported JSONL source is unchanged
- Only a `TEMP` table was created and modified

## Reproduction
```bash
printf 'CREATE TEMP TABLE scratch(id INTEGER);\nINSERT INTO scratch VALUES (1);\n.save --force\n' | sqly testdata/sample.jsonl
```

## Actual Behavior
`sqly` tries to save the untouched JSONL import and fails with `write-back to sample.jsonl is not supported`.

## Expected Behavior
No save attempt should happen because no file-backed imported table changed.

## Suggested Regression Coverage
- Add an e2e test for `.save --force` after temp-table-only changes on a JSONL import.
- Assert that the session stays successful and does not attempt any write-back.

### LINKED ISSUE #533: [BUG] interactive .save DIR fails on an unchanged JSONL import after only TEMP-table changes
## Trigger Conditions
- Published `v0.22.0` binary
- Interactive `.save DIR`
- Imported JSONL source is unchanged
- Only a `TEMP` table was created and modified

## Reproduction
```bash
printf 'CREATE TEMP TABLE scratch(id INTEGER);\nINSERT INTO scratch VALUES (1);\n.save /tmp/jsonl-temp-only-out\n' | sqly testdata/sample.jsonl
```

## Actual Behavior
`sqly` tries to export the untouched JSONL import and fails with `write-back to sample.jsonl is not supported`.

## Expected Behavior
No save/export attempt should happen because no file-backed imported table changed.

## Suggested Regression Coverage
- Add an e2e test for `.save DIR` after temp-table-only changes on a JSONL import.
- Assert that the command succeeds without creating output files or write-back errors.

### LINKED ISSUE #534: [BUG] interactive .save --force fails when only a SQL-created scratch table changed
## Trigger Conditions
- Published `v0.22.0` binary
- Interactive `.save --force`
- Imported CSV source is unchanged
- Only a SQL-created non-file-backed table was modified

## Reproduction
```bash
cp testdata/sample.csv /tmp/scratch_only.csv
printf 'CREATE TABLE scratch(id INTEGER);\nINSERT INTO scratch VALUES (1);\n.save --force\n' | sqly /tmp/scratch_only.csv
```

## Actual Behavior
`sqly` aborts with `scratch: not loaded from a file` even though the imported CSV table never changed.

## Expected Behavior
The save should ignore non-file-backed scratch tables when deciding whether imported sources need to be written.

## Suggested Regression Coverage
- Add an e2e test for `.save --force` after changes limited to a SQL-created scratch table.
- Assert that the unchanged imported source is left alone and no `not loaded from a file` error appears.

### LINKED ISSUE #535: [BUG] interactive .save DIR fails when only a SQL-created scratch table changed
## Trigger Conditions
- Published `v0.22.0` binary
- Interactive `.save DIR`
- Imported CSV source is unchanged
- Only a SQL-created non-file-backed table was modified

## Reproduction
```bash
cp testdata/sample.csv /tmp/scratch_only_dir.csv
printf 'CREATE TABLE scratch(id INTEGER);\nINSERT INTO scratch VALUES (1);\n.save /tmp/scratch-only-out\n' | sqly /tmp/scratch_only_dir.csv
```

## Actual Behavior
`sqly` aborts with `scratch: not loaded from a file` even though the imported CSV table never changed.

## Expected Behavior
The export should ignore non-file-backed scratch tables when deciding whether imported sources need to be written.

## Suggested Regression Coverage
- Add an e2e test for `.save DIR` after changes limited to a SQL-created scratch table.
- Assert that no output directory is populated and no `not loaded from a file` error appears.

### LINKED ISSUE #536: [BUG] interactive .save --force reports unrelated JSONL write-back errors when only a scratch table changed
## Trigger Conditions
- Published `v0.22.0` binary
- Interactive `.save --force`
- Imported JSONL source is unchanged
- Only a SQL-created non-file-backed table was modified

## Reproduction
```bash
printf 'CREATE TABLE scratch(id INTEGER);\nINSERT INTO scratch VALUES (1);\n.save --force\n' | sqly testdata/sample.jsonl
```

## Actual Behavior
`sqly` reports both `sample: write-back to sample.jsonl is not supported` and `scratch: not loaded from a file`, even though neither condition should matter when the imported table is unchanged.

## Expected Behavior
No save attempt should happen; the command should ignore both the untouched JSONL import and the unrelated scratch table.

## Suggested Regression Coverage
- Add an e2e test for `.save --force` after scratch-table-only changes on a JSONL import.
- Assert that no write-back or scratch-table errors are reported.

### LINKED ISSUE #537: [BUG] interactive .save DIR reports unrelated JSONL write-back errors when only a scratch table changed
## Trigger Conditions
- Published `v0.22.0` binary
- Interactive `.save DIR`
- Imported JSONL source is unchanged
- Only a SQL-created non-file-backed table was modified

## Reproduction
```bash
printf 'CREATE TABLE scratch(id INTEGER);\nINSERT INTO scratch VALUES (1);\n.save /tmp/jsonl-scratch-out\n' | sqly testdata/sample.jsonl
```

## Actual Behavior
`sqly` reports both `sample: write-back to sample.jsonl is not supported` and `scratch: not loaded from a file`, even though neither condition should matter when the imported table is unchanged.

## Expected Behavior
No export attempt should happen; the command should ignore both the untouched JSONL import and the unrelated scratch table.

## Suggested Regression Coverage
- Add an e2e test for `.save DIR` after scratch-table-only changes on a JSONL import.
- Assert that no output files are written and no unrelated errors are reported.

### LINKED ISSUE #538: [BUG] interactive .save --force rewrites the source after net-zero CSV updates
## Trigger Conditions
- Published `v0.22.0` binary
- Interactive `.save --force`
- An imported CSV table is updated and then restored to its original contents

## Reproduction
```bash
cp testdata/sample.csv /tmp/netzero.csv
printf 'UPDATE netzero SET age=99 WHERE id=1;\nUPDATE netzero SET age=41 WHERE id=1;\n.save --force\n' | sqly /tmp/netzero.csv
```

## Actual Behavior
`sqly` prints `Saved netzero to /tmp/netzero.csv` and rewrites the file bytes even though the final relational contents match the original input.

## Expected Behavior
A net-zero session should not trigger write-back when the imported table content is unchanged at save time.

## Suggested Regression Coverage
- Add an e2e test that performs net-zero updates before `.save --force`.
- Assert that the original file bytes remain unchanged and no save message is emitted.

### LINKED ISSUE #539: [BUG] interactive .save DIR exports after net-zero CSV updates
## Trigger Conditions
- Published `v0.22.0` binary
- Interactive `.save DIR`
- An imported CSV table is updated and then restored to its original contents

## Reproduction
```bash
cp testdata/sample.csv /tmp/netzero_dir.csv
rm -rf /tmp/netzero-out
printf 'UPDATE netzero_dir SET age=99 WHERE id=1;\nUPDATE netzero_dir SET age=41 WHERE id=1;\n.save /tmp/netzero-out\n' | sqly /tmp/netzero_dir.csv
```

## Actual Behavior
`sqly` writes `/tmp/netzero-out/netzero_dir.csv` even though the final relational contents match the original input.

## Expected Behavior
A net-zero session should not trigger an export when the imported table content is unchanged at save time.

## Suggested Regression Coverage
- Add an e2e test that performs net-zero updates before `.save DIR`.
- Assert that no output file is created when the imported data matches the original snapshot.

### LINKED ISSUE #540: [BUG] --sql-file --save --force rewrites the source after net-zero CSV updates
## Trigger Conditions
- Published `v0.22.0` binary
- Non-interactive `--sql-file --save --force`
- An imported CSV table is updated and then restored to its original contents

## Reproduction
```bash
cp testdata/sample.csv /tmp/netzero_file.csv
printf 'UPDATE netzero_file SET age=99 WHERE id=1;\nUPDATE netzero_file SET age=41 WHERE id=1;\n' > /tmp/netzero.sql
sqly --sql-file /tmp/netzero.sql --save --force /tmp/netzero_file.csv
```

## Actual Behavior
`sqly` rewrites `/tmp/netzero_file.csv` even though the final relational contents match the original input.

## Expected Behavior
A net-zero batch session should not trigger write-back when the imported table content is unchanged at the end of the script.

## Suggested Regression Coverage
- Add an e2e test for `--sql-file --save --force` with net-zero updates.
- Assert that the file bytes stay unchanged and no save confirmation is emitted.

### LINKED ISSUE #541: [BUG] --sql-file --save-dir exports after net-zero CSV updates
## Trigger Conditions
- Published `v0.22.0` binary
- Non-interactive `--sql-file --save-dir`
- An imported CSV table is updated and then restored to its original contents

## Reproduction
```bash
cp testdata/sample.csv /tmp/netzero_dir_file.csv
printf 'UPDATE netzero_dir_file SET age=99 WHERE id=1;\nUPDATE netzero_dir_file SET age=41 WHERE id=1;\n' > /tmp/netzero_dir.sql
sqly --sql-file /tmp/netzero_dir.sql --save-dir /tmp/netzero-batch-out /tmp/netzero_dir_file.csv
```

## Actual Behavior
`sqly` writes `/tmp/netzero-batch-out/netzero_dir_file.csv` even though the final relational contents match the original input.

## Expected Behavior
A net-zero batch session should not trigger an export when the imported table content is unchanged at the end of the script.

## Suggested Regression Coverage
- Add an e2e test for `--sql-file --save-dir` with net-zero updates.
- Assert that no output file is created when the imported data matches the original snapshot.

### PR DESCRIPTION
**Pull request by nao1215**
## Summary

Write-back (`.save`, `.save DIR`, `--save`/`--save-dir`) keyed off a single session-wide `dataChanged` flag set whenever any DML affected a row, including DML on a TEMP table or a SQL-created scratch table. So a session that changed no imported source still drove write-back over the untouched imports:

- TEMP-table-only or scratch-table-only changes rewrote an unchanged CSV (#530, #531), failed on an unchanged JSONL import (#532, #533, #536, #537), or aborted on the scratch table itself with `not loaded from a file` (#534, #535).
- Net-zero edits that cancel out (an UPDATE later reversed) rewrote a source whose content matched the original input (#538-#541).

## Fix

Record a content fingerprint for each imported table at load time. At save time, write back only a file-backed table whose current content differs from its baseline:

- A non-file-backed scratch table is skipped (it has no source file to persist).
- An unchanged import is skipped before the writability check, so an untouched JSONL/Excel import never reports "not supported".
- A net-zero edit leaves the source byte-identical and emits no save line.
- After a successful write the baseline moves forward, so a later `.save` in the same session is a no-op.

Preflight validation is unchanged: it still validates every file-backed table up front (the unchanged-skip is disabled there, since at preflight no change has happened yet), so a genuinely unwritable change still fails loudly.

## Tests

Go: `TestSaveCommandSkipsWhenNoImportedTableChanged` (TEMP-only, scratch-only, net-zero) and `TestSaveCommandSkipsUnwritableImportWhenUnchanged` (unchanged JSONL while a scratch table changed). `TestSaveCommandPersistsAfterDataChange` and the directory-import rejection tests still guard that genuine changes save and that a changed directory import is still rejected. Verified the new tests fail on `main` before the fix.

shellspec: `spec/v0_22_bugs_spec.sh` runs the built binary for TEMP-only, scratch-only, net-zero (interactive and `--sql-file`), and a genuine-change control that still saves.

Closes #530
Closes #531
Closes #532
Closes #533
Closes #534
Closes #535
Closes #536
Closes #537
Closes #538
Closes #539
Closes #540
Closes #541

## Summary by CodeRabbit

* **Bug Fixes**
  * `.save`, `.save DIR`, and `--save`/`--save-dir` now only persist imported tables when their content actually changes
  * Sessions modifying only TEMP or scratch tables, or with net-zero edits, no longer trigger unnecessary rewrites or errors
  * A "nothing to save" message displays when no imported tables have changed
  * Unwritable imports (e.g., JSONL) that remain unchanged are skipped without generating errors

### CODE REVIEW

File: shell/import.go [hunk 1]
```diff
@@ -595,9 +598,72 @@ func (s *Shell) recordTableSources(tableNames []string, source string) {
...
+	for _, col := range t.Header() {
+		writeField(col)
+	}
+	for _, rec := range t.Records() {
+		writeField("\x00") // row separator that no column value can forge
+		for _, f := range rec {
+			writeField(f)
+		}
+	}
+	return hex.EncodeToString(h.Sum(nil)), nil
```
[the remark this answers is not in the document]
Comment (nao1215) [reply shell/import.go:648]: Thanks, but sorting the rows here would be incorrect for this design. The write path that actually persists the table (executeWriteBack in save.go) reads it with the same s.usecases.metadata.List call (SELECT *) and hands those rows straight to DumpTable, so the file is written in whatever order List returns. The fingerprint reads through the exact same List call, so it already mirrors the bytes that would be written: "fingerprint unchanged" is equivalent to "the file we would write is byte-identical to the baseline".

If we sorted rows only in the fingerprint, a session that legitimately reorders rows (for example a delete + reinsert that lands new rowids) would fingerprint as unchanged and we would skip a save that does change the output file, dropping the new ordering. For the reported net-zero cases (UPDATE then UPDATE back) the rowid is stable, so List returns the same order at import and at save and the fingerprint matches correctly. A single in-memory connection scanning an unmodified table returns rows in stable rowid order, so there is no nondeterminism between the import-time and save-time reads of an unchanged table. Keeping the fingerprint in List order (rather than a sorted canonical order) is what guarantees skip-iff-identical-output, so I am leaving it as is.
